### PR TITLE
Rating opt in

### PR DESCRIPTION
**Pull request by gestchild**
References #4411 

Allows people to opt in to rating results

optin

### CODE REVIEW

File: common/views/components/OptIn/OptIn.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    optedInMessage: string[],
+    optInCTA: string,
+    optOutCTA: string,
+  |},
+  cookieName: string,
+|};
+
+const OptIn = ({ text, cookieName }: Props) => {
+  const [shouldRender, setShouldRender] = useState(false);
+  const [cookieValue, setCookieValue] = useState(false);
```
Comment (jamesgorrie) [inline common/views/components/OptIn/OptIn.js]: Is this more about the active state rather than cookie? The cookie is just a means of setting / getting that, and it's not really the cookie value (one being Boolean the other string).

File: common/views/components/OptIn/OptIn.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  }
+`;
+
+const HiddenText = styled.span.attrs({
+  className: 'visually-hidden',
+})``;
+
+type Props = {|
+  text: {|
+    defaultMessage: string[],
```
Comment (jamesgorrie) [inline common/views/components/OptIn/OptIn.js]: Misunderstood what this meant, normally default meaning fallback.
Is this the callout?

File: common/views/components/OptIn/OptIn.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    setCookieValue(true);
+    updateToggles({
+      [`${cookieName}`]: true,
+    });
+  }
+
+  useEffect(() => {
+    const cookieValue = cookie.get(`toggle_${cookieName}`);
+    setCookieValue(cookieValue);
```
Comment (jamesgorrie) [inline common/views/components/OptIn/OptIn.js]: We're mixing boolean and string values to this state. We should probably type our states.

File: common/views/pages/_app.js [hunk 4]
```diff
@@ -153,9 +157,18 @@ export default class WecoApp extends App {
...
     }
+
     super(props);
   }
 
+  state: State = {
+    togglesContext: toggles,
+  };
+
+  updateToggles = (newToggles: Object) => {
```
Comment (jamesgorrie) [inline common/views/pages/_app.js]: We're moving the functionality of the toggles into even more places here.
The component is setting the cookie, here it's updating the context, and it's being read by some middleware.

I wonder if we create a toggles context that can do all of this in one place, simplifying the component reading and writing to this.

This works - and happy to move this conversation into another PR.

File: common/views/components/OptIn/OptIn.js (hunk 2 above)
Comment (gestchild) [reply common/views/components/OptIn/OptIn.js]: The message changes depending on whether your opted in or not.
From this: 
Screenshot 2019-10-31 at 16 01 33

To this:
Screenshot 2019-10-31 at 16 01 50

defaultMessage is the one you get if you're not opted in. Would you rather I call it calloutMessage?

File: common/views/pages/_app.js (hunk 4 above)
Comment (gestchild) [reply common/views/pages/_app.js]: 👍

File: common/views/components/OptIn/OptIn.js (hunk 2 above)
Comment (jamesgorrie) [reply common/views/components/OptIn/OptIn.js]: What about `optInMessage` and `optedInMessage`?

Comment (jamesgorrie) [reply common/views/components/OptIn/OptIn.js]: Bikeshedding question around whether we need this abstracted if we're just using it in one place, and we can think about the abstraction when we have more use cases.

File: common/views/components/OptIn/OptIn.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    boolean,
+    ((boolean => boolean) | boolean) => void
+  ] = useState(false);
+  const [optedIn, setOptedIn]: [
+    boolean,
+    ((boolean => boolean) | boolean) => void
+  ] = useState(false);
+  const { updateToggles } = useContext(TogglesContext);
+  function optOut() {
+    cookie.set(`toggle_${cookieName}`, 'false', {
```
Comment (jamesgorrie) [inline common/views/components/OptIn/OptIn.js]: Could `updateToggles` not do this?

File: common/views/components/OptIn/OptIn.js (hunk 2 above)
Comment (gestchild) [reply common/views/components/OptIn/OptIn.js]: I'll make it static for now, then don't have to worry about the naming

File: common/views/components/OptIn/OptIn.js (hunk 5 above)
Comment (gestchild) [reply common/views/components/OptIn/OptIn.js]: Probably, might leave as is for now and do it when looking at the  toggles context in general

File: catalogue/webapp/pages/works.js [hunk 6]
```diff
@@ -339,8 +353,9 @@ const Works = ({ works, searchParams }: Props) => {
                         />
                       </div>
                       <TogglesContext.Consumer>
-                        {({ relevanceRating }) =>
-                          relevanceRating && (
+                        {({ relevanceRating, relevanceRatingOptIn }) =>
```
Comment (jamesgorrie) [inline catalogue/webapp/pages/works.js]: What's the difference between these two? Should we not just have 1?

Comment (gestchild) [reply catalogue/webapp/pages/works.js]: One is for the actual ratings (relevanceRating), the other is because Jenn wanted the whole thing behind a toggle so she could take a look at it all first
